src/shared/install.c

```diff
@@ -577,20 +578,29 @@ static int create_symlink(
          * right place, or negative on error.
          */
 
-        (void) mkdir_parents_label(new_path, 0755);
+        if (!dry_run)
+                (void) mkdir_parents_label(new_path, 0755);
 
-        if (symlink(old_path, new_path) >= 0) {
+        if (!dry_run && symlink(old_path, new_path) >= 0) {
                 r = install_changes_add(changes, n_changes, INSTALL_CHANGE_SYMLINK, new_path, old_path);
                 if (r < 0)
                         return r;
                 return 1;
         }
 
-        if (errno != EEXIST)
+        if (!dry_run && errno != EEXIST)
                 return install_changes_add(changes, n_changes, -errno, new_path, NULL);
 
-        r = readlink_malloc(new_path, &dest);
+        /* For dry-run, we need to check if symlink exists. For regular operation, we reach here due to EEXIST */
+        if (dry_run)
+                r = is_symlink(new_path) > 0 ? readlink_malloc(new_path, &dest) : -ENOENT;
```

Comment:
This is complicated, and hard to understand.
Why not to check it earlier around `symlink()` is called?